Entities/ACDropShip.cpp

```diff
@@ -982,206 +774,123 @@ void ACDropShip::Update()
 */
 }
 
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          ResetEmissionTimers
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Reset the timers of all emissions so they will start/stop at the 
-//                  correct relative offsets from now.
-
-void ACDropShip::ResetEmissionTimers()
-{
-    if (m_pRThruster && m_pRThruster->IsAttached())
-        m_pRThruster->ResetEmissionTimers();
-
-    if (m_pLThruster && m_pLThruster->IsAttached())
-        m_pLThruster->ResetEmissionTimers();
-
-    if (m_pURThruster && m_pURThruster->IsAttached())
-        m_pURThruster->ResetEmissionTimers();
-
-    if (m_pULThruster && m_pULThruster->IsAttached())
-        m_pULThruster->ResetEmissionTimers();
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void ACDropShip::SetRightThruster(Attachable *newThruster) {
+    if (newThruster == nullptr) {
+        if (m_pRThruster && m_pRThruster->IsAttachedTo(this)) { RemoveAttachable(m_pRThruster); }
+        m_pRThruster = nullptr;
+    } else {
+        AEmitter *castedNewThruster = dynamic_cast<AEmitter *>(newThruster);
+        if (castedNewThruster) {
+            RemoveAttachable(m_pRThruster);
```

Comment:
Hmmm, 2 points to make:
1. Regarding the missing check:
To be honest the check in the 1st branch of the if-else is kind of a leftover that I didn't think about much; in previous versions of the code things would have ended up in really weird states or endless loops without it, but I added more check to RemoveAttachable over time so it's not as necessary as it once was. That said, it's still a little helpful cause it eliminates a superfluous function call.
To walk through things step-by-step, say I lose a thruster from impulse forces: this calls RemoveAttachable(thruster) which removes the attachable, sets its parent to null and calls SetThruster(nullptr). If I don't have the check there, it'll call RemoveAttachable again, which'll early return cause the Attachable has no parent anymore, but it ends up being a bit of a waste.
As for options, I don't want to have an IsAttachedTo check in the 2nd branch of the if-else, cause RemoveAttachable has an assert that triggers for this case, and I want that, cause it should never happen and if it does, we should know.
I could replace IsAttachedTo with IsAttached and have essentially the same benefit without messing with the assert, or I could get rid of it and accept the superfluous function call.

2. Handling the case where thruster existing but is not attached to this
This should trigger the assert in RemoveAttachable??

Reply:
I should've written more clearly, the case that I meant is where the thruster exists and is not attached to anything.
I had not considered that it could exist and be attached to something else.
Both of my questions still stand:
1. If you want to prevent an unnecessary function call, why not do the same check (exists and attached to something) inside the else-block?
2. What happens to the thruster object if it exists but is not attached to anything when this function is called? In this case the RemoveAttachable assert will not trigger.

Reply:
I should've written more clearly, the case that I meant is where the thruster exists and is not attached to anything.
I had not considered that it could exist and be attached to something else.
But now from your example I understand that you assume that if the thruster isn't attached to anything then RemoveAttachable was already called on it.
And in that case the thruster object is taken care of, and if it needs to be deleted then it will be deleted.
So that takes care of my second question - something will take care of deleting the old thruster object.

For the first point, as you've said, you can do the same check (exists and attached to something) inside the else-block as well.
Not that it matters so much, just seems odd to only do it on the if and not on the else.

Reply:
Yeah, I did it in the 1st branch of the if-else because it was needed at one point. You're right though that there's no reason not to do it in the 2nd branch too. I'll add it to all of them.